### DISCUSSION TITLE
uBlock and CORS error logs in Firefox

### DISCUSSION THREAD

**Discussion by reallyuniquename** [uBlockOrigin/uBlock-issues#3420] [Q&A] [answered]
When uBlock blocks requests to some domain Firefox generates error logs in browser console such as

```
Cross-Origin Request Blocked: The Same Origin Policy disallows reading the remote resource at https://manyads.com/metric/58hf91f. (Reason: CORS request did not succeed). Status code: (null).
```

I understand it is related to uBlock which is confirmed by network tab that explicitly states a request was blocked by uBlock.

The issue is it floods the console in dev tools especially if some ad script keeps requesting a page every second. Most importantly it makes it hard to read genuine CORS errors or any javascript errors at all.

Is there a way to disable these?

**Comment by stephenhawk8054** [comment]
Maybe disable `redirect-rule` for XHR?

```adb
@@*$xhr,redirect-rule
```

**Reply by reallyuniquename** [reply]
I don't think it has anything to do with that. Even a `<script src="...">` request which is blocked by a dynamic filter for a whole domain still generates the same error on Firefox. 

Maybe uBlock somehow injects custom CSP to block network requests? I honestly don't know. I checked the issue tracker and it seems other users get the same error https://github.com/uBlockOrigin/uBlock-issues/issues/478 and even use it to check whether something is actually blocked.

**Reply by stephenhawk8054** [reply]
Can you give example site?

**Reply by reallyuniquename** [reply]
Sure, `github.com`. Create a dynamic blocking filter for `collector.github.com`, open Firefox dev tools console, refresh page. You should see CORS errors.

**Reply by stephenhawk8054** [reply]
I don't see CORS error in `github.com`

![image](https://github.com/user-attachments/assets/047c7e0e-53c4-4ccc-bab3-734945e24596)

**Reply by reallyuniquename** [reply]
Sorry, wrong example. Go to `https://symless.com/synergy`, create a dynamic filter for `play.google.com` or a static filter `||play.google.com/log`, open Firefox console, refresh page.

**Reply by stephenhawk8054** [reply]
It's Firefox issue with `XHR`. Even when I don't use any extensions and set Firefox Enhanced Tracking Protection to `Strict`, Firefox Strict mode blocks the XHR connection `https://googleads.g.doubleclick.net/pagead/id` by its own and the console shows the same CORS error too

![image](https://github.com/user-attachments/assets/07378124-9c7b-47cc-83c7-7d33894ef6bd)

**Reply by stephenhawk8054** [reply]
So regardless of which tool you use, not necessarily uBO, if it blocks XHR, it could show errors. Maybe not always, but it's not a uBO specific issue.

**Reply by garry-ut99** [reply]
Similiar to (maybe related or even maybe a duplicate?):
- https://github.com/uBlockOrigin/uBlock-issues/issues/1222

**Reply by reallyuniquename** [reply]
@stephenhawk8054 I double checked everything and you are right. Any blocking of a cross domain XHR would generate such errors in the logs. Any ideas how to suppress them if requests are blocked by uBlock?

What initially threw me off is in some cases `<script src="...">` also produces same message in the console. It turns out if a page embeds a cross domain script dynamically into the DOM it also counts as XHR. Here's a snippet that emulates that.

```js
          d = window.document;
          h = d.getElementsByTagName('head')[0];
          s = d.createElement('script');
          s.crossOrigin = 'anonymous';
          s.async = 1;
          s.src = 'https://cdn.jetboost.io/jetboost.js';
          h.appendChild(s);
```

**Reply by stephenhawk8054** [reply]
> Any ideas how to suppress them if requests are blocked by uBlock?

You'll need to use scriptlet filters which trap `XMLHttpRequest` or `fetch` like `##+js(prevent-xhr)` and `##+js(prevent-fetch)`

For example, in `https://symless.com/synergy` page:

```adb
youtube.com##+js(prevent-xhr, play.google.com/log)
```

will not generate errors in the console.

---

> It turns out if a page embeds a cross domain script dynamically into the DOM it also counts as XHR.

I don't remember I've seen a case like this. Can you find a site that has this issue?

**Reply by reallyuniquename** [reply]
Any way to default to `prevent-xhr`? Otherwise it would cover just my own filters.

>Can you find a site that has this issue?

Open `https://github.com/uBlockOrigin/uBlock-issues`, block `githubassets.com`, refresh page. You should see a lot of CORS errors which are not exactly XHR related. 

```js
<script crossorigin="anonymous" defer="defer" type="application/javascript" src="https://github.githubassets.com/assets/wp-runtime-e655dbaff571.js"></script>
```

Notice that blocked scripts are statically embeded with `crossOrigin` attribute yet they produce the same error. I guess there is no need for  extra scripts to reproduce this.

**Reply by gorhill** [reply, marked as answer]
> Any way to default to `prevent-xhr`?

No, `prevent-xhr` is scriptlet injection, there is no universal way to do this for all sites.

Best you bring your issue to Firefox, it's where something can be done if they choose to.

**Reply by stephenhawk8054** [reply]
Also from the [MDN documentation](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS/Errors/CORSDidNotSucceed), the error is simply `CORS request did not succeed` so it could simply mean the connection fails to fire when you block them in general, not necessarily related to CORS.

> The [HTTP](https://developer.mozilla.org/en-US/docs/Glossary/HTTP) request which makes use of CORS failed because the HTTP connection failed at either the network or protocol level. The error is not directly related to CORS, but is a fundamental network error of some kind.
> 
> In many cases, it is caused by a browser plugin (e.g. an ad blocker or privacy protector) blocking the request.

It's different to other [CORS Errors reasons](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS/Errors#cors_error_messages).
